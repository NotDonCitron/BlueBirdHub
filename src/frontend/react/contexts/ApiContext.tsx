import React, { createContext, useContext, useState, ReactNode } from 'react';
import { API_CONFIG, getApiUrl } from '../config/api';

type ApiStatus = 'connected' | 'disconnected' | 'checking';

interface ApiContextType {
  apiStatus: ApiStatus;
  setApiStatus: (status: ApiStatus) => void;
  retryConnection: () => void;
  makeApiRequest: (endpoint: string, method?: string, data?: any, headers?: Record<string, string>) => Promise<any>;
}

export const ApiContext = createContext<ApiContextType | undefined>(undefined);

interface ApiProviderProps {
  children: ReactNode;
  apiStatus: ApiStatus;
  onRetryConnection: () => void;
}

export const ApiProvider: React.FC<ApiProviderProps> = ({ 
  children, 
  apiStatus: initialStatus, 
  onRetryConnection 
}) => {
  const [apiStatus, setApiStatus] = useState<ApiStatus>(initialStatus);

  const retryConnection = () => {
    setApiStatus('checking');
    onRetryConnection();
  };

  const makeApiRequest = async (endpoint: string, method = 'GET', data?: any, headers?: Record<string, string>): Promise<any> => {
    try {
      // Get auth token from localStorage
      const authToken = localStorage.getItem('auth_token');
      const authHeaders = authToken ? { 'Authorization': `Bearer ${authToken}` } : {};
      
      if ((window as any).electronAPI) {
        // Use Electron API
        return await (window as any).electronAPI.apiRequest(endpoint, method, data, { ...authHeaders, ...headers });
      } else {
        // Use configured API URL (works in both dev and production)
        const url = getApiUrl(endpoint);
        const options: RequestInit = {
          method,
          headers: {
            'Content-Type': 'application/json',
            ...authHeaders,
            ...headers
          },
        };
        
        if (data && method !== 'GET') {
          options.body = JSON.stringify(data);
        }
        
        const response = await fetch(url, options);
        if (!response.ok) {
          throw new Error(`HTTP error! status: ${response.status}`);
        }
        
        return await response.json();
      }
    } catch (error) {
      console.error('API request failed:', error);
      setApiStatus('disconnected');
      throw error;
    }
  };

  const value: ApiContextType = {
    apiStatus,
    setApiStatus,
    retryConnection,
    makeApiRequest,
  };

  return (
    <ApiContext.Provider value={value}>
      {children}
    </ApiContext.Provider>
  );
};

export const useApi = (): ApiContextType => {
  const context = useContext(ApiContext);
  if (context === undefined) {
    throw new Error('useApi must be used within an ApiProvider');
  }
  return context;
};