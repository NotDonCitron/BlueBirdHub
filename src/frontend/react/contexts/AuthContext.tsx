import React, { createContext, useContext, useState, useEffect, ReactNode } from 'react';

interface User {
  id: number;
  username: string;
  last_login: string | null;
  created_at: string;
}

interface AuthContextType {
  token: string | null;
  user: User | null;
  isAuthenticated: boolean;
  isLoading: boolean;
  login: (token: string) => Promise<void>;
  logout: () => void;
}

const AuthContext = createContext<AuthContextType | undefined>(undefined);

interface AuthProviderProps {
  children: ReactNode;
}

export const AuthProvider: React.FC<AuthProviderProps> = ({ children }) => {
  const [token, setToken] = useState<string | null>(null);
  const [user, setUser] = useState<User | null>(null);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    // Check for stored token on mount  
    const storedToken = localStorage.getItem('auth_token');
    if (storedToken) {
      setToken(storedToken);
      fetchUserInfo(storedToken);
    } else {
      setIsLoading(false);
    }
  }, []);

  const fetchUserInfo = async (authToken: string) => {
    try {
      const headers = {
        'Authorization': `Bearer ${authToken}`
      };

      let response;
      if ((window as any).electronAPI) {
        response = await (window as any).electronAPI.apiRequest('/auth/me', 'GET', null, headers);
      } else {
        const res = await fetch('http://127.0.0.1:8000/auth/me', {
          headers
        });
        if (!res.ok) {
          throw new Error('Failed to fetch user info');
        }
        response = await res.json();
      }

      setUser(response.user);
    } catch (error) {
      console.error('Failed to fetch user info:', error);
      // If fetching user info fails, clear the token
      logout();
    } finally {
      setIsLoading(false);
    }
  };

  const login = async (authToken: string) => {
    setToken(authToken);
    localStorage.setItem('auth_token', authToken);
    await fetchUserInfo(authToken);
  };

  const logout = async () => {
    // Inform backend about logout
    if (token) {
      try {
        const headers = {
          'Authorization': `Bearer ${token}`
        };

        if ((window as any).electronAPI) {
          await (window as any).electronAPI.apiRequest('/auth/logout', 'POST', null, headers);
        } else {
          await fetch('http://127.0.0.1:8000/auth/logout', {
            method: 'POST',
            headers
          });
        }
      } catch (error) {
        console.error('Failed to logout on backend:', error);
        // Continue with frontend logout even if backend fails
      }
    }

    // Clear frontend state
    setToken(null);
    setUser(null);
    localStorage.removeItem('auth_token');
  };

  const value = {
    token,
    user,
    isAuthenticated: !!token && !!user,
    isLoading,
    login,
    logout
  };

  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

export const useAuth = () => {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
};