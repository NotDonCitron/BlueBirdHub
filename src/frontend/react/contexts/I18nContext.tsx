import React, { createContext, useContext, useEffect, useState, ReactNode } from 'react';
import { useTranslation } from 'react-i18next';
import { 
  supportedLanguages, 
  rtlLanguages, 
  getCurrentLanguage, 
  changeLanguage, 
  getAvailableLanguages,
  formatCurrency,
  formatDate,
  formatRelativeTime,
  preloadNamespace
} from '../config/i18n';

// Types
export interface Language {
  code: string;
  name: string;
  nativeName: string;
  rtl: boolean;
}

export interface I18nContextType {
  // Current language state
  currentLanguage: string;
  isRTL: boolean;
  availableLanguages: Language[];
  
  // Language management
  changeLanguage: (lng: string) => Promise<void>;
  getCurrentLanguage: () => string;
  
  // Translation functions
  t: (key: string, options?: any) => string;
  
  // Formatting functions
  formatCurrency: (amount: number, currency?: string) => string;
  formatDate: (date: Date | string, style?: 'short' | 'medium' | 'long' | 'full') => string;
  formatRelativeTime: (date: Date | string) => string;
  formatNumber: (number: number, options?: Intl.NumberFormatOptions) => string;
  
  // Namespace management
  loadNamespace: (ns: string) => Promise<void>;
  
  // Pluralization helpers
  getPlural: (key: string, count: number, options?: any) => string;
  
  // Validation
  isKeyExists: (key: string, ns?: string) => boolean;
  
  // Gender support
  getGenderedText: (key: string, gender: 'male' | 'female' | 'neutral', options?: any) => string;
  
  // Loading state
  isLoading: boolean;
  
  // Error handling
  translationError: string | null;
}

const I18nContext = createContext<I18nContextType | undefined>(undefined);

interface I18nProviderProps {
  children: ReactNode;
  defaultLanguage?: string;
}

export const I18nProvider: React.FC<I18nProviderProps> = ({ 
  children, 
  defaultLanguage 
}) => {
  const { t, i18n, ready } = useTranslation();
  const [currentLanguage, setCurrentLanguage] = useState<string>(getCurrentLanguage());
  const [isRTL, setIsRTL] = useState<boolean>(rtlLanguages.includes(getCurrentLanguage()));
  const [isLoading, setIsLoading] = useState<boolean>(!ready);
  const [translationError, setTranslationError] = useState<string | null>(null);
  
  const availableLanguages = getAvailableLanguages();

  // Language change handler
  const handleLanguageChange = async (lng: string): Promise<void> => {
    try {
      setIsLoading(true);
      setTranslationError(null);
      
      // Validate language
      if (!supportedLanguages[lng as keyof typeof supportedLanguages]) {
        throw new Error(`Unsupported language: ${lng}`);
      }
      
      // Change language
      await changeLanguage(lng);
      
      // Update state
      setCurrentLanguage(lng);
      setIsRTL(rtlLanguages.includes(lng));
      
      // Update document attributes
      document.documentElement.lang = lng;
      document.documentElement.dir = rtlLanguages.includes(lng) ? 'rtl' : 'ltr';
      
      // Store preference
      localStorage.setItem('preferredLanguage', lng);
      
    } catch (error) {
      setTranslationError(`Failed to change language: ${error}`);
      console.error('Language change error:', error);
    } finally {
      setIsLoading(false);
    }
  };

  // Format number with current locale
  const formatNumber = (number: number, options: Intl.NumberFormatOptions = {}): string => {
    try {
      return new Intl.NumberFormat(currentLanguage, options).format(number);
    } catch (error) {
      console.warn('Number formatting error:', error);
      return number.toString();
    }
  };

  // Load additional namespace
  const loadNamespace = async (ns: string): Promise<void> => {
    try {
      setIsLoading(true);
      await preloadNamespace(ns, currentLanguage);
    } catch (error) {
      setTranslationError(`Failed to load namespace: ${ns}`);
      console.error('Namespace loading error:', error);
    } finally {
      setIsLoading(false);
    }
  };

  // Get plural form
  const getPlural = (key: string, count: number, options: any = {}): string => {
    return t(key, { count, ...options });
  };

  // Check if translation key exists
  const isKeyExists = (key: string, ns?: string): boolean => {
    return i18n.exists(key, { ns });
  };

  // Get gendered text
  const getGenderedText = (key: string, gender: 'male' | 'female' | 'neutral', options: any = {}): string => {
    const genderedKey = `${key}_${gender}`;
    if (isKeyExists(genderedKey)) {
      return t(genderedKey, options);
    }
    // Fallback to neutral or base key
    return t(key, options);
  };

  // Listen for language changes
  useEffect(() => {
    const handleI18nLanguageChange = (lng: string) => {
      setCurrentLanguage(lng);
      setIsRTL(rtlLanguages.includes(lng));
    };

    i18n.on('languageChanged', handleI18nLanguageChange);
    
    return () => {
      i18n.off('languageChanged', handleI18nLanguageChange);
    };
  }, [i18n]);

  // Initialize default language
  useEffect(() => {
    if (defaultLanguage && defaultLanguage !== currentLanguage) {
      handleLanguageChange(defaultLanguage);
    }
  }, [defaultLanguage]);

  // Listen for i18n ready state
  useEffect(() => {
    setIsLoading(!ready);
  }, [ready]);

  // Context value
  const contextValue: I18nContextType = {
    currentLanguage,
    isRTL,
    availableLanguages,
    changeLanguage: handleLanguageChange,
    getCurrentLanguage: () => currentLanguage,
    t,
    formatCurrency,
    formatDate,
    formatRelativeTime,
    formatNumber,
    loadNamespace,
    getPlural,
    isKeyExists,
    getGenderedText,
    isLoading,
    translationError
  };

  return (
    <I18nContext.Provider value={contextValue}>
      {children}
    </I18nContext.Provider>
  );
};

// Custom hook to use I18n context
export const useI18n = (): I18nContextType => {
  const context = useContext(I18nContext);
  if (context === undefined) {
    throw new Error('useI18n must be used within an I18nProvider');
  }
  return context;
};

// HOC for class components
export const withI18n = <P extends object>(Component: React.ComponentType<P>) => {
  return React.forwardRef<any, P>((props, ref) => {
    const i18nContext = useI18n();
    return <Component {...props} ref={ref} i18n={i18nContext} />;
  });
};

// Custom hooks for specific use cases
export const useTranslationWithNamespace = (namespace: string) => {
  const { t, loadNamespace } = useI18n();
  
  useEffect(() => {
    loadNamespace(namespace);
  }, [namespace, loadNamespace]);
  
  return (key: string, options?: any) => t(`${namespace}:${key}`, options);
};

export const useRTL = () => {
  const { isRTL } = useI18n();
  return isRTL;
};

export const useLanguageDirection = () => {
  const { isRTL } = useI18n();
  return isRTL ? 'rtl' : 'ltr';
};

export const useLocalizedFormatting = () => {
  const { formatCurrency, formatDate, formatRelativeTime, formatNumber } = useI18n();
  
  return {
    currency: formatCurrency,
    date: formatDate,
    relativeTime: formatRelativeTime,
    number: formatNumber
  };
};

export default I18nContext;