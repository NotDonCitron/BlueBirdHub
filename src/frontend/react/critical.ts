// Critical path resources for immediate rendering
// This module contains only the essential code needed for the initial page render

// Critical CSS imports - these will be inlined
import './styles/critical.css';

// Critical components for above-the-fold content
export { default as Header } from './components/Header/Header';
export { default as Layout } from './components/Layout/Layout';
export { default as LoadingSkeleton } from './components/LoadingSkeleton/LoadingSkeleton';

// Essential utilities for initial render
export { default as ErrorBoundary } from './components/ErrorBoundary';

// Critical context providers
export { ApiProvider } from './contexts/ApiContext';
export { ThemeProvider } from './contexts/ThemeContext';