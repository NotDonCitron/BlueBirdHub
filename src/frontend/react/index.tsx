import React from 'react';
import { createRoot } from 'react-dom/client';
import './utils/consoleFilter'; // Apply console filtering for browser extensions
import ErrorBoundary from './components/ErrorBoundary';
import './styles/global.css';

// Import the full app with error boundary protection
const SimpleApp = React.lazy(() => import('./SimpleApp'));

// Loading component
const LoadingScreen: React.FC = () => (
  <div style={{ 
    display: 'flex', 
    justifyContent: 'center', 
    alignItems: 'center', 
    height: '100vh',
    backgroundColor: '#f0f0f0',
    fontFamily: 'Arial, sans-serif'
  }}>
    <div style={{ textAlign: 'center' }}>
      <div style={{ 
        fontSize: '48px', 
        marginBottom: '20px',
        animation: 'spin 2s linear infinite'
      }}>
        ðŸ§ 
      </div>
      <h2 style={{ color: '#4a9eff', marginBottom: '10px' }}>Loading BlueBirdHub...</h2>
      <p style={{ color: '#666' }}>Initializing AI Agent System</p>
      <style dangerouslySetInnerHTML={{
        __html: `
          @keyframes spin {
            from { transform: rotate(0deg); }
            to { transform: rotate(360deg); }
          }
        `
      }} />
    </div>
  </div>
);

// Main App Component with Suspense
const App: React.FC = () => (
  <ErrorBoundary>
    <React.Suspense fallback={<LoadingScreen />}>
      <SimpleApp />
    </React.Suspense>
  </ErrorBoundary>
);

// Get the root element
const container = document.getElementById('root');
if (!container) {
  throw new Error('Failed to find the root element');
}

// Create root and render app
const root = createRoot(container);
root.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>
);