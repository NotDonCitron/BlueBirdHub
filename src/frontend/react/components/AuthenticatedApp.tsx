import React, { Suspense } from 'react';
import { LazyLayout, LazyErrorDashboard, LazyPerformanceDashboard } from './LazyComponents';
import LoadingSkeleton from './LoadingSkeleton/LoadingSkeleton';
import { useAuth } from '../contexts/AuthContext';
import Login from './Login/Login';

const AuthenticatedApp: React.FC = () => {
  const { isAuthenticated, isLoading, login } = useAuth();

  if (isLoading) {
    return (
      <div className="app-loading">
        <div className="loading-spinner"></div>
        <p>Loading...</p>
      </div>
    );
  }

  if (!isAuthenticated) {
    return <Login onLogin={login} />;
  }

  return (
    <Suspense fallback={<LoadingSkeleton />}>
      <LazyLayout />
      <LazyErrorDashboard />
      <LazyPerformanceDashboard />
    </Suspense>
  );
};

export default AuthenticatedApp;