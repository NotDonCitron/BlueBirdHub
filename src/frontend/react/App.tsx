import React, { useState, useEffect, useCallback } from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import AuthenticatedApp from './components/AuthenticatedApp';
import ErrorBoundary from './components/ErrorBoundary';
import { ThemeProvider } from './contexts/ThemeContext';
import { ApiProvider } from './contexts/ApiContext';
import { AuthProvider } from './contexts/AuthContext';
import { performanceMonitor } from './utils/performanceMonitor';
import { resourcePreloader } from './utils/resourcePreloader';
import { preloadCriticalComponents } from './components/LazyComponents';
import './styles/App.css';

const App: React.FC = () => {
  const [isLoading, setIsLoading] = useState(true);
  const [apiStatus, setApiStatus] = useState<'connected' | 'disconnected' | 'checking'>('checking');

  const initializeApp = useCallback(async () => {
    // Start performance monitoring
    performanceMonitor.startContinuousMonitoring();
    
    // Preload critical resources
    const resourceResults = await resourcePreloader.preloadCriticalResources();
    console.log('Resource preload results:', resourceResults);
    
    // Preload critical components
    preloadCriticalComponents();
    
    // Check API connection
    await checkApiConnection();
  }, []);

  useEffect(() => {
    initializeApp();
    
    return () => {
      performanceMonitor.dispose();
    };
  }, [initializeApp]);

  const checkApiConnection = useCallback(async () => {
    try {
      // Add global object for browser compatibility
  (window as any).global = window;
  
  // Check if we're running in Electron
  if ((window as any).electronAPI) {
        const response = await (window as any).electronAPI.apiRequest('/', 'GET');
        if (response.status === 'running') {
          setApiStatus('connected');
        } else {
          setApiStatus('disconnected');
        }
      } else {
        // Fallback for development in browser
        const response = await fetch('http://127.0.0.1:8000/');
        const data = await response.json();
        if (data.status === 'running') {
          setApiStatus('connected');
        } else {
          setApiStatus('disconnected');
        }
      }
    } catch (error) {
      console.error('Failed to connect to API:', error);
      setApiStatus('disconnected');
    } finally {
      setIsLoading(false);
    }
  }, []);

  if (isLoading) {
    return (
      <div className="app-loading">
        <div className="loading-spinner"></div>
        <p>Initializing OrdnungsHub...</p>
      </div>
    );
  }

  return (
    <ErrorBoundary>
      <Router>
        <ThemeProvider>
          <ApiProvider apiStatus={apiStatus} onRetryConnection={checkApiConnection}>
            <AuthProvider>
              <div className="app">
                <AuthenticatedApp />
              </div>
            </AuthProvider>
          </ApiProvider>
        </ThemeProvider>
      </Router>
    </ErrorBoundary>
  );
};

export default App;