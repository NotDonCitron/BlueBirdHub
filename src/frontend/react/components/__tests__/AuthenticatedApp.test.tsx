import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import AuthenticatedApp from '../AuthenticatedApp';
import { useAuth } from '../../contexts/AuthContext';

// Mock the dependencies
jest.mock('../../contexts/AuthContext');
jest.mock('../Layout/Layout', () => {
  return function MockLayout() {
    return <div data-testid="layout">Layout Component</div>;
  };
});
jest.mock('../ErrorDashboard', () => {
  return function MockErrorDashboard() {
    return <div data-testid="error-dashboard">Error Dashboard</div>;
  };
});
jest.mock('../PerformanceDashboard', () => {
  return function MockPerformanceDashboard() {
    return <div data-testid="performance-dashboard">Performance Dashboard</div>;
  };
});
jest.mock('../Login/Login', () => {
  return function MockLogin({ onLogin }: { onLogin: (token: string) => void }) {
    const handleLogin = () => {
      onLogin('test-token');
    };
    
    return (
      <div data-testid="login">
        <div>Login Component</div>
        <button onClick={handleLogin}>Mock Login</button>
      </div>
    );
  };
});

const mockUseAuth = useAuth as jest.MockedFunction<typeof useAuth>;

describe('AuthenticatedApp Component', () => {
  const mockLogin = jest.fn();

  const defaultAuthContext = {
    token: 'test-token',
    user: {
      id: 1,
      username: 'testuser',
      email: 'test@example.com',
      is_active: true,
    },
    isAuthenticated: true,
    isLoading: false,
    login: mockLogin,
    logout: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('Loading State', () => {
    it('renders loading spinner when isLoading is true', () => {
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        isLoading: true,
        isAuthenticated: false,
        user: null,
        token: null,
      });

      render(<AuthenticatedApp />);

      expect(screen.getByText('Loading...')).toBeInTheDocument();
      
      // Should have loading spinner element
      const loadingContainer = screen.getByText('Loading...').parentElement;
      expect(loadingContainer).toHaveClass('app-loading');
      expect(loadingContainer?.querySelector('.loading-spinner')).toBeInTheDocument();
    });

    it('does not render other components when loading', () => {
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        isLoading: true,
        isAuthenticated: false,
        user: null,
        token: null,
      });

      render(<AuthenticatedApp />);

      expect(screen.queryByTestId('layout')).not.toBeInTheDocument();
      expect(screen.queryByTestId('login')).not.toBeInTheDocument();
      expect(screen.queryByTestId('error-dashboard')).not.toBeInTheDocument();
      expect(screen.queryByTestId('performance-dashboard')).not.toBeInTheDocument();
    });
  });

  describe('Authentication States', () => {
    it('renders Login component when not authenticated and not loading', () => {
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        isAuthenticated: false,
        user: null,
        token: null,
        isLoading: false,
      });

      render(<AuthenticatedApp />);

      expect(screen.getByTestId('login')).toBeInTheDocument();
      expect(screen.getByText('Login Component')).toBeInTheDocument();
      
      // Should not render authenticated components
      expect(screen.queryByTestId('layout')).not.toBeInTheDocument();
      expect(screen.queryByTestId('error-dashboard')).not.toBeInTheDocument();
      expect(screen.queryByTestId('performance-dashboard')).not.toBeInTheDocument();
    });

    it('passes login function to Login component', async () => {
      const user = userEvent.setup();
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        isAuthenticated: false,
        user: null,
        token: null,
        isLoading: false,
      });

      render(<AuthenticatedApp />);

      const loginButton = screen.getByText('Mock Login');
      await user.click(loginButton);

      expect(mockLogin).toHaveBeenCalledWith('test-token');
    });

    it('renders authenticated components when authenticated', () => {
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        isAuthenticated: true,
      });

      render(<AuthenticatedApp />);

      expect(screen.getByTestId('layout')).toBeInTheDocument();
      expect(screen.getByTestId('error-dashboard')).toBeInTheDocument();
      expect(screen.getByTestId('performance-dashboard')).toBeInTheDocument();
      
      // Should not render login or loading
      expect(screen.queryByTestId('login')).not.toBeInTheDocument();
      expect(screen.queryByText('Loading...')).not.toBeInTheDocument();
    });
  });

  describe('Component Integration', () => {
    it('renders all authenticated components together', () => {
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        isAuthenticated: true,
      });

      render(<AuthenticatedApp />);

      expect(screen.getByText('Layout Component')).toBeInTheDocument();
      expect(screen.getByText('Error Dashboard')).toBeInTheDocument();
      expect(screen.getByText('Performance Dashboard')).toBeInTheDocument();
    });

    it('renders components in correct order', () => {
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        isAuthenticated: true,
      });

      render(<AuthenticatedApp />);

      // All components should be present
      expect(screen.getByTestId('layout')).toBeInTheDocument();
      expect(screen.getByTestId('error-dashboard')).toBeInTheDocument();
      expect(screen.getByTestId('performance-dashboard')).toBeInTheDocument();
    });
  });

  describe('State Transitions', () => {
    it('transitions from loading to login when authentication fails', () => {
      // First render with loading state
      const { rerender } = render(<AuthenticatedApp />);
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        isLoading: true,
        isAuthenticated: false,
        user: null,
        token: null,
      });

      rerender(<AuthenticatedApp />);
      expect(screen.getByText('Loading...')).toBeInTheDocument();

      // Then transition to login state
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        isLoading: false,
        isAuthenticated: false,
        user: null,
        token: null,
      });

      rerender(<AuthenticatedApp />);
      expect(screen.queryByText('Loading...')).not.toBeInTheDocument();
      expect(screen.getByTestId('login')).toBeInTheDocument();
    });

    it('transitions from loading to authenticated when authentication succeeds', () => {
      // First render with loading state
      const { rerender } = render(<AuthenticatedApp />);
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        isLoading: true,
        isAuthenticated: false,
        user: null,
        token: null,
      });

      rerender(<AuthenticatedApp />);
      expect(screen.getByText('Loading...')).toBeInTheDocument();

      // Then transition to authenticated state
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        isLoading: false,
        isAuthenticated: true,
      });

      rerender(<AuthenticatedApp />);
      expect(screen.queryByText('Loading...')).not.toBeInTheDocument();
      expect(screen.getByTestId('layout')).toBeInTheDocument();
    });

    it('transitions from login to authenticated after successful login', () => {
      // First render with login state
      const { rerender } = render(<AuthenticatedApp />);
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        isLoading: false,
        isAuthenticated: false,
        user: null,
        token: null,
      });

      rerender(<AuthenticatedApp />);
      expect(screen.getByTestId('login')).toBeInTheDocument();

      // Then transition to authenticated state
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        isLoading: false,
        isAuthenticated: true,
      });

      rerender(<AuthenticatedApp />);
      expect(screen.queryByTestId('login')).not.toBeInTheDocument();
      expect(screen.getByTestId('layout')).toBeInTheDocument();
    });

    it('transitions from authenticated to login when logged out', () => {
      // First render with authenticated state
      const { rerender } = render(<AuthenticatedApp />);
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        isLoading: false,
        isAuthenticated: true,
      });

      rerender(<AuthenticatedApp />);
      expect(screen.getByTestId('layout')).toBeInTheDocument();

      // Then transition to login state
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        isLoading: false,
        isAuthenticated: false,
        user: null,
        token: null,
      });

      rerender(<AuthenticatedApp />);
      expect(screen.queryByTestId('layout')).not.toBeInTheDocument();
      expect(screen.getByTestId('login')).toBeInTheDocument();
    });
  });

  describe('Edge Cases', () => {
    it('handles undefined user in authenticated state', () => {
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        user: undefined as any,
        isAuthenticated: true,
        isLoading: false,
      });

      render(<AuthenticatedApp />);

      expect(screen.getByTestId('layout')).toBeInTheDocument();
      expect(screen.getByTestId('error-dashboard')).toBeInTheDocument();
      expect(screen.getByTestId('performance-dashboard')).toBeInTheDocument();
    });

    it('handles null token in authenticated state', () => {
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        token: null,
        isAuthenticated: true,
        isLoading: false,
      });

      render(<AuthenticatedApp />);

      expect(screen.getByTestId('layout')).toBeInTheDocument();
      expect(screen.getByTestId('error-dashboard')).toBeInTheDocument();
      expect(screen.getByTestId('performance-dashboard')).toBeInTheDocument();
    });

    it('prioritizes loading state over authentication state', () => {
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        isLoading: true,
        isAuthenticated: true, // This should be ignored when loading
      });

      render(<AuthenticatedApp />);

      expect(screen.getByText('Loading...')).toBeInTheDocument();
      expect(screen.queryByTestId('layout')).not.toBeInTheDocument();
      expect(screen.queryByTestId('login')).not.toBeInTheDocument();
    });

    it('handles loading with existing user data', () => {
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        isLoading: true,
        isAuthenticated: false, // Not yet confirmed
        user: {
          id: 1,
          username: 'testuser',
          email: 'test@example.com',
          is_active: true,
        },
        token: 'existing-token',
      });

      render(<AuthenticatedApp />);

      expect(screen.getByText('Loading...')).toBeInTheDocument();
      expect(screen.queryByTestId('layout')).not.toBeInTheDocument();
      expect(screen.queryByTestId('login')).not.toBeInTheDocument();
    });
  });

  describe('Accessibility', () => {
    it('provides accessible loading indicator', () => {
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        isLoading: true,
        isAuthenticated: false,
        user: null,
        token: null,
      });

      render(<AuthenticatedApp />);

      const loadingText = screen.getByText('Loading...');
      expect(loadingText).toBeInTheDocument();
      
      // The loading container should be properly structured
      const loadingContainer = loadingText.parentElement;
      expect(loadingContainer).toHaveClass('app-loading');
    });

    it('maintains proper semantic structure for authenticated state', () => {
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        isAuthenticated: true,
      });

      const { container } = render(<AuthenticatedApp />);

      // The main content should be wrapped in a React Fragment
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('CSS Classes', () => {
    it('applies correct CSS class to loading container', () => {
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        isLoading: true,
        isAuthenticated: false,
        user: null,
        token: null,
      });

      render(<AuthenticatedApp />);

      const loadingContainer = screen.getByText('Loading...').parentElement;
      expect(loadingContainer).toHaveClass('app-loading');
    });

    it('applies correct CSS class to loading spinner', () => {
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        isLoading: true,
        isAuthenticated: false,
        user: null,
        token: null,
      });

      render(<AuthenticatedApp />);

      const loadingContainer = screen.getByText('Loading...').parentElement;
      const spinner = loadingContainer?.querySelector('.loading-spinner');
      expect(spinner).toBeInTheDocument();
      expect(spinner).toHaveClass('loading-spinner');
    });
  });

  describe('Component Props', () => {
    it('passes correct onLogin prop to Login component', () => {
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        isAuthenticated: false,
        user: null,
        token: null,
        isLoading: false,
      });

      render(<AuthenticatedApp />);

      // The Login component should receive the login function
      expect(screen.getByTestId('login')).toBeInTheDocument();
      
      // This is tested indirectly through the mock Login component
      // which calls onLogin when the Mock Login button is clicked
    });
  });

  describe('Performance', () => {
    it('does not render unnecessary components when loading', () => {
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        isLoading: true,
        isAuthenticated: false,
        user: null,
        token: null,
      });

      render(<AuthenticatedApp />);

      // Only loading should be rendered
      expect(screen.getByText('Loading...')).toBeInTheDocument();
      
      // These should not be in the DOM at all
      expect(screen.queryByTestId('layout')).not.toBeInTheDocument();
      expect(screen.queryByTestId('login')).not.toBeInTheDocument();
      expect(screen.queryByTestId('error-dashboard')).not.toBeInTheDocument();
      expect(screen.queryByTestId('performance-dashboard')).not.toBeInTheDocument();
    });

    it('does not render authenticated components when not authenticated', () => {
      mockUseAuth.mockReturnValue({
        ...defaultAuthContext,
        isAuthenticated: false,
        user: null,
        token: null,
        isLoading: false,
      });

      render(<AuthenticatedApp />);

      // Only login should be rendered
      expect(screen.getByTestId('login')).toBeInTheDocument();
      
      // These should not be in the DOM at all
      expect(screen.queryByTestId('layout')).not.toBeInTheDocument();
      expect(screen.queryByTestId('error-dashboard')).not.toBeInTheDocument();
      expect(screen.queryByTestId('performance-dashboard')).not.toBeInTheDocument();
    });
  });
});