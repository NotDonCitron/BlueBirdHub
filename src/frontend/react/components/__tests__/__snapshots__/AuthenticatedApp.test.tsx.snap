// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthenticatedApp Component Accessibility maintains proper semantic structure for authenticated state 1`] = `
<div
  data-testid="layout"
>
  Layout Component
</div>
`;
